src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java

```diff
@@ -0,0 +1,42 @@
+package org.hydev.mcpm.client.database.searchusecase;
+
+import org.hydev.mcpm.client.models.PluginModel;
+import org.hydev.mcpm.client.models.PluginVersion;
+
+import java.util.*;
+
+/**
+ * Searcher that returns a map based on keywords.
+ *
+ * @author Jerry Zhu (<a href="https://github.com/jerryzhu509">...</a>)
+ */
+public class SearcherByKeyword implements Searcher {
+
+    /**
+     * Returns a dictionary mapping the different keywords to the matching plugins
+     * This function should ignore letter case in searching.
+     * For example, searching "java" would match "Java" as well.
+     * This function implements fuzzy search that doesn't require the exact phrase to be available,
+     * but requires all words in the phrase to be present.
+     * For example, searching "java 11" would match "java jdk 11" but not "java"
```

Comment:
The memory complexity required for fuzzy searching would be infinite given the way the code is currently set up. 
At the moment, you pass the index all the way to the controller.
However, fuzzy searching requires extra work to be done during runtime, and it cannot be precomputed (you can still precompute things but you need to do extra work). 
I suggest that instead of having your searchers return an index, they take in a the parameters and return the search result. This way you can implement fuzzy searching in the inner layer instead of the outer layer, which is not supposed to deal with these things.

Reply:
I don't fully understand. Wdm by "pass the index all the way to the controller".
Searching by keywords just basically returns the plugins whose descriptions contain all the words within the "keyword" string, the essence of which is the keywords to list of plugins map. This is created in the business rules layer as far as i can tell

Reply:
Right now, you create a mapping from single keywords to plugins.
However, this map is given to the database interactor.
The database interactor now has to do some logical work to determine which plugins have *all* the keywords in the search query, unless your map can take in every possible query and output a plugin for that.
The point is that all the work should be done in the innermost part in clean architecture, and the database interactor shouldn't have to worry about it, but this should be the job of the searcher.

Reply:
Yeah the point of the searchers is to do the hard work in computing the plugins for *every* keyword. All the database interactor does in searching is a dictionary look up basically

Reply:
It only does every singular keyword. Currently it cannot do multiple keywords.

Suppose I did wanted to add a function to search plugins that have every keyword in my query.

One way I could do this, through clean architecture, is the following:

1. Add a new search type to the enum and create the corresponding searcher that implements the Searcher interface.
2. Implement the interface: I need to return a map. This needs to be a mapping from multiple keywords to a list of plugins. I don't really want to change the Searcher interface, which requires a mapping from String to list, but that's okay, we can represent multiple keywords as a string separated with spaces. No problem yet.
3. Now I need to go through every plugin, and see what possible combination of keywords could lead me to this plugin. Thankfully, if the search query contains a keyword not in my plugin's keywords, this plugin shouldn't be included. So we only consider keywords in the plugin.
4. Now I just need to add this plugin to every possible combination of keywords. Well, if I have N keywords, then every possible combination would be 2^N - 1 (we exclude the empty set). However, I only counted the subsets. Subsets are not ordered, but input queries can be in any order! This means I need to take every subset and also add the plugin to every permutation of it, so I can cover every query.
5. We went through every plugin and can return the map.

Now let's look at what happens above, and try to change parts that look like it took too much work to do something.

1. Step 1 is ok.
2. I could possibly change the interface to take in a set of strings.
3. This is ok.
4. Hm, input queries are in any order. If I had the query with me, I could sort all the keywords. This would solve my issue of subsets being unordered. Or, if I had the query with me, I could actually just use a mapping from single keywords to a list of plugins: The answer would be the intersection of the lists of plugins matching each keyword. Ideally I would also change the mapping to use sets, so I can speed up the intersection.
5. In the previous step I just computed the answer, so we can return the answer instead.

This is the fundamental problem: most the hard work is done by the searcher. All the database interactor does is look up the answer. But what if we get a new problem where we can't look up the answer (unless we sacrifice something)? Then we have to move all the work to the searcher, the searcher can give the answer to the database interactor, and then for the specific case where you need to do a little extra, you can add it to the lower level class instead of the higher level class.